
# Principal Component Analysis (PCA)


*   PCA dengan data Iris
*   Perbandingan PCA dengan model KNN
*   Perbandingan PCA dengan model Naive Bayes


Sumber [dataset](https://en.wikipedia.org/wiki/Iris_flower_data_set) 


Install Package yang diperlukan

In [2]:
!pip install scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


Impor _package_ yang digunakan

In [79]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier# Create KNN classifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import decomposition
from sklearn import datasets
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split #split dataset into train and test data
from sklearn.metrics import accuracy_score



*   Ambil data _iris_ dari _Sklearn datasets library_ 
*   Pisahkan data ```X``` dengan kelas ```y```



In [49]:
# load features and targets separately
iris = datasets.load_iris()
X = iris.data
y = iris.target

Normalisasi dengan ```MinMaxScaler```

In [50]:
scaler = MinMaxScaler()
scala = scaler.fit(X)

In [51]:
normalized_dataset = scaler.transform(X)

_# data latih (nilai data)_

```X_train```

_# data tes (nilai data)_

```X_test```

_# data latih (kelas data)_

```y_train```

_# data tes (kelas data)_

```y_test```

In [52]:
# percent_amount_of_test_data = / HUNDRED_PERCENT
percent_amount_of_test_data = 0.2

In [53]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = percent_amount_of_test_data, random_state=1, stratify=y)

## Model PCA

In [54]:
from sklearn.decomposition import PCA

pca = PCA()
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

### 3 Kolom

In [65]:
from sklearn.decomposition import PCA
 
pca = PCA(n_components=1)
 
pca_features = pca.fit_transform(normalized_dataset)
 
print('Dimensi sebelum PCA: ', normalized_dataset.shape)
print('Dimensi setelah PCA: ', pca_features.shape)
 
pca_df = pd.DataFrame(
    data=pca_features, 
    columns=['PC1'])

Dimensi sebelum PCA:  (150, 4)
Dimensi setelah PCA:  (150, 1)


In [66]:
target_names = {
    0:'setosa',
    1:'versicolor', 
    2:'virginica'
}
 
pca_df['target'] = y
pca_df['target'] = pca_df['target'].map(target_names)
 
pca_df.head(5)

PC1  target
0 -0.630703  setosa
1 -0.622905  setosa
2 -0.669520  setosa
3 -0.654153  setosa
4 -0.648788  setosa

In [70]:
# nilai variance dari setiap ratio
explained_variance = pca.explained_variance_ratio_
print(explained_variance)

[0.84136038]


In [67]:
classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

In [68]:
cm = confusion_matrix(y_test, y_pred)
print('Akurasi = ' + str(accuracy_score(y_test, y_pred)))

Akurasi = 0.9333333333333333


## Model KNN

In [72]:
amount_of_neighbor = 3

In [73]:
# Fit the classifier to the data
knn = KNeighborsClassifier(n_neighbors = amount_of_neighbor)
# create train data
knn.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=3)

In [75]:
knn.predict(X_test)

array([2, 0, 1, 0, 0, 0, 2, 2, 2, 1, 0, 1, 2, 1, 2, 0, 2, 1, 1, 2, 1, 1,
       0, 0, 2, 1, 0, 0, 1, 1])

In [76]:
#check accuracy of our model on the test data
knn.score(X_test, y_test)

0.9666666666666667

In [80]:
# classifier
gaussian = GaussianNB()

gaussian.fit(X_train, y_train)
Y_pred = gaussian.predict(X_test) 
accuracy_nb=round(accuracy_score(y_test,Y_pred)* 100, 2)
acc_gaussian = round(gaussian.score(X_train, y_train) * 100, 2)

accuracy = accuracy_score(y_test,Y_pred)
print('accuracy_Naive Bayes: %.3f' %accuracy)

accuracy_Naive Bayes: 0.900


## Kesimpulan

*   Akurasi model KNN = $0.967$
*   Akurasi model PCA (Kolom: 1/2/3) = $0.934$
*   Akurasi model Naive Bayes = $0.9$

